freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/EnvironmentUserSyncStateCalculator.java

```diff
@@ -0,0 +1,93 @@
+        String envCrnString = environmentCrn.toString();
+
+        Stack stack = stackService.getByEnvironmentCrnAndAccountId(envCrnString, accountId);
+        UserSyncStatus userSyncStatus = userSyncStatusService.findByStack(stack);
+
+        return internalCalculateEnvironmentUserSyncState(accountId, envCrnString, userSyncStatus);
+    }
+
+    @VisibleForTesting
+    EnvironmentUserSyncState internalCalculateEnvironmentUserSyncState(String accountId, String envCrnString, UserSyncStatus userSyncStatus) {
+        EnvironmentUserSyncState environmentUserSyncState = new EnvironmentUserSyncState();
+        if (null == userSyncStatus || null == userSyncStatus.getLastStartedFullSync()) {
+            environmentUserSyncState.setState(UserSyncState.STALE);
+        } else {
+            environmentUserSyncState.setLastUserSyncOperationId(userSyncStatus.getLastStartedFullSync().getOperationId());
+            environmentUserSyncState.setState(calculateUserSyncState(accountId, envCrnString, userSyncStatus));
+        }
+
+        return environmentUserSyncState;
+    }
+
+    private UserSyncState calculateUserSyncState(String accountId, String envCrnString, UserSyncStatus userSyncStatus) {
+        Operation lastSync = userSyncStatus.getLastStartedFullSync();
+        UserSyncState state = UserSyncState.STALE;
+        switch (lastSync.getStatus()) {
+            case RUNNING:
+                state = UserSyncState.SYNCING;
+                break;
+            case COMPLETED:
+                if (lastSync.getSuccessList().stream()
+                        .anyMatch(details -> envCrnString.equals(details.getEnvironment()))) {
+                    UmsEventGenerationIds currentEventGenerationIds = umsEventGenerationIdsProvider.getEventGenerationIds(accountId, MDCUtils.getRequestId());
+                    if (eventGenerationIdsChecker.isInSync(userSyncStatus, currentEventGenerationIds)) {
+                        state = UserSyncState.IN_SYNC;
+                    }
+                }
+                break;
+            case REQUESTED:
+            case REJECTED:
+                // REQUESTED or REJECTED operations will never be saved as part of the UserSyncStatus
+                LOGGER.error("UserSyncStatus.lastStartedFullSync '{}' in unexpected state {} for environment '{}'",
+                        lastSync, lastSync.getStatus(), envCrnString);
+                throw new IllegalStateException(
+                        String.format("Last sync operation [%s] for environment '%s' is in unexpected state %s",
+                                lastSync.getOperationId(), envCrnString, lastSync.getStatus()));
+            case TIMEDOUT:
+                LOGGER.warn("UserSyncStatus.lastStartedFullSync '{}' is timed out for environment '{}'", lastSync.getOperationId(), envCrnString);
+                break;
+            case FAILED:
+            default:
+                // state already set to STALE
```

Comment:
it would be easier to read if every case would set the state instead of we initialize with a value and we have to go through all case to check if it's overridden